language-plutus-core/src/Language/PlutusCore/MkPlc.hs

```diff
@@ -66,6 +71,30 @@ class TermLike term tyname name uni | term -> tyname, term -> name, term -> uni
     termLet  :: ann -> TermDef term tyname name uni ann -> term ann -> term ann
     typeLet  :: ann -> TypeDef tyname uni ann -> term ann -> term ann
 
+-- | A class for computing annotations from Haskell values.
+class ToAnnotation uni ann where
```

Comment:
This module used to just be convenience things, but now it's morphing into an abstraction over various kinds of terms, it seems? Maybe that's fine?

I also wonder if we should bundle a few of these classes together into one with some associated types, e.g. we could have
```
class TermLike .... where
    UniOf term :: * -> *
```
etc.?

Reply:
> This module used to just be convenience things, but now it's morphing into an abstraction over various kinds of terms, it seems? Maybe that's fine?

This is just a first approximation of the whole structure. We'll need to decide whether Untyped Plutus Core should depend on Typed Plutus Core or they should depend on something third.

> I also wonder if we should bundle a few of these classes together into one with some associated types, e.g. we could have

Untyped `Term` is not going to be `TermLike`, because the latter contains stuff like `error` receiving a type, `iWrap`, `typeLet`, etc. `TermLike` is only meant to be used for typed things and we need `UniOf` for untyped ones as well.